test/passing/gen/gen.ml

```diff
@@ -0,0 +1,72 @@
+module StringMap = Map.Make (String)
+
+type setup =
+  { mutable has_ref: bool
+  ; mutable has_opts: bool
+  ; mutable base_file: string option
+  ; mutable extra_deps: string list }
+
+let read_lines fn =
+  Stdio.In_channel.with_file fn ~f:Stdio.In_channel.input_lines
+
+let add_test ?base_file map src_test_name =
+  let s = {has_ref= false; has_opts= false; base_file; extra_deps= []} in
+  map := StringMap.add src_test_name s !map ;
+  s
+
+let register_file tests fname =
+  match String.split_on_char '.' fname with
+  | test_name :: (("ml" | "mli" | "mlt") as ext) :: rest -> (
+      let src_test_name = test_name ^ "." ^ ext in
+      let setup =
+        match StringMap.find src_test_name !tests with
+        | setup -> setup
+        | exception Not_found -> (
+          (* foo_file-some_variant.ml should derive from foo_file.ml *)
+          match String.index_opt test_name '-' with
+          | None -> add_test tests src_test_name
+          | Some i ->
+              let base_file = String.sub test_name 0 i ^ "." ^ ext in
+              add_test ~base_file tests src_test_name )
+      in
+      match rest with
+      | [] -> ()
+      | ["output"] -> ()
+      | ["opts"] -> setup.has_opts <- true
+      | ["ref"] -> setup.has_ref <- true
+      | ["deps"] -> setup.extra_deps <- read_lines fname
+      | _ -> invalid_arg fname )
+  | _ -> ()
+
+(* ignore dune file, .foo.whatever.swp, etc *)
+
+let emit_test test_name setup =
+  let open Printf in
+  let opts =
+    if setup.has_opts then sprintf " %%{read-lines:%s.opts}" test_name
+    else ""
+  in
+  let ref_name = if setup.has_ref then test_name ^ ".ref" else test_name in
+  let base_test_name =
+    match setup.base_file with Some n -> n | None -> test_name
+  in
+  let extra_deps = String.concat " " setup.extra_deps in
+  Printf.printf
+    {|
+(rule
+ (targets %s.output)
+ (deps .ocamlformat %s)
+ (action
+   (with-outputs-to %%{targets}
+     (system "%%{bin:ocamlformat}%s %%{dep:%s} || true"))))
```

Comment:
Unrelated but `true` won't work on Windows.
Also, I think we should not ignore errors here. Then the diff between the whole file and the error message is shown and can be promoted, which is not useful and looks bad.

Reply:
Ah yes I forgot to mention that. Some tests do fail, though. What do you think about the following (in a separate PR):

- by default we generate `(run %{bin:ocamlformat} ...)` (fails on error)
- if the `TEST.should-fail` file exists, this is converted to `(system "! ocamlformat ...")` (fails on success

We can convert that to `with-accepted-exit-codes` in dune 2.

Reply:
Yes that's a good idea.